### PR TITLE
[FIX] sale_coupon: Allow coupons to correctly allow several free products

### PR DESCRIPTION
**Pull request by abz-odoo**
What are the steps to reproduce your issue ?

    1. Add a coupon program with the following condition sale_ok == True & name == 'Large Desk'
    2. Add 'Buy 1 get 2 free' 'Large Cabinet'
    3. Create a quotation with 1 Large Desk and 2 Large Cabinet
    4. Add the coupon code

What is currently happening ?

    Promotion will only remove one * Large Cabinet price instead of 2

What are you expecting to happen ?

    Promotion remove two * Large Cabinet price

Why is this happening ?

    Because there is a filter that applies before the calculation of the number of free articles obtained is done.
    This filter only keeps the products that match the condition of the coupon: the product that must be purchased in order to benefit two free items.
    In our example this filter just keeps Large Desk, which is the item you have to buy to get 2 Large Cabinet.
    The calculation is not done correctly because it only has 1 Large Desk,
    and will apply the promotion for 1 Large Cabinet

How to fix the bug ?

    Remove this first filter

opw-2496940

### CODE REVIEW

File: addons/sale_coupon/models/sale_order.py [hunk 1]
```diff
@@ -45,6 +45,7 @@ def copy(self, default=None):
     def action_confirm(self):
         self.generated_coupon_ids.write({'state': 'new'})
         self.applied_coupon_ids.write({'state': 'used'})
+        self.recompute_coupon_lines()
```
Comment (agr-odoo) [inline addons/sale_coupon/models/sale_order.py:48]: why is this necessary?

**Review by agr-odoo — COMMENTED** [review]
Looks ok beside a small doubt (see below)

File: addons/sale_coupon/models/sale_order.py (hunk 1 above)
Comment (abz-odoo) [reply addons/sale_coupon/models/sale_order.py:48]: because the coupon is not automatically updated each time we modify the quantity (linked to your message on the previous PR) suddenly I force the recalculation of the coupon each time the quotation is confirmed to have the right numbers

it is not the case in 12.0, only in 13.0 and 14.0

Comment (agr-odoo) [reply addons/sale_coupon/models/sale_order.py:48]: Even if this seems odd it is the current behavior of the module now ;)
https://github.com/odoo/enterprise/commit/5f6371ee6ade2a29dac8acccf60dc8d65d2afec1

**Comment by agr-odoo** [comment]
ok now
robodoo delegate+

**Comment by abz-odoo** [comment]
thank you

robodoo r+
